llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -3731,8 +3731,10 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
   }
   case ISD::SUB: {
     EVT VT = Node->getValueType(0);
-    assert(TLI.isOperationLegalOrCustom(ISD::ADD, VT) &&
-           TLI.isOperationLegalOrCustom(ISD::XOR, VT) &&
+    assert((VT.isFixedLengthVector() || // fixed length ADD can be expanded to
```

Comment:
This change looks weird.  The assert is checking whether it is safe to create `ADD` and `XOR` nodes, which you're bypassing?  What does the failing test case look like?  Part of me thinks the original assert should be an `if` statement, but then I'm also wondering why `VectorLegalizer::LegalizeOp` hasn't already unrolled the operation.

Reply:
That's right. The way the code now works (if you'd ignore the assert) is that it expands a `sub` into a `add` + `xor`. If those are vector operations that also require expansion, they can be expanded further (in a subsequent step) into scalar operations. The assert is trying to match the way the code currently works, which allows the code to be successfully compiled. I figured updating the assert was the most sensible thing to do for now, given that it's not really a common use-case to optimise for.

Reply:
To me the assert exists to ensure optimal code, which is to say it doesn't really make sense to convert the sub to an inverted add unless both those operations are in someway legal.  When not the better result would be to just expand the sub.  Which I guess makes the assert a bear trap for the first person that needs the support.